renovate-aarch32/src/Renovate/Arch/AArch32/ISA.hs

```diff
@@ -512,143 +507,144 @@ markLast lst =
+  -> MD.ParsedBlock MA.ARM ids
+  -> R.ConcreteAddress MA.ARM
+  -> Instruction tp ()
+  -> Maybe (R.Relocation MA.ARM)
+jumpTargetRelocation mem toSymbolic pb insnAddr i =
+  case armJumpType i mem insnAddr pb of
+    Some (R.RelativeJump _jc src offset) -> Just (R.SymbolicRelocation (toSymbolic (src `R.addressAddOffset` offset)))
+    Some (R.AbsoluteJump _jc target) -> Just (R.SymbolicRelocation (toSymbolic target))
+    Some (R.IndirectJump {}) -> Nothing
+    Some (R.DirectCall src offset) -> Just (R.SymbolicRelocation (toSymbolic (src `R.addressAddOffset` offset)))
+    Some R.IndirectCall -> Nothing
+    Some (R.Return {}) -> Nothing
+    Some R.NoJump -> Nothing
+    Some (R.NotInstrumentable _) -> Nothing
+
+noRelocation :: DA.Operand tp -> DA.Annotated (R.Relocation MA.ARM) DA.Operand tp
+noRelocation = DA.Annotated R.NoRelocation
+
 armSymbolizeAddresses :: MM.Memory 32
+                      -> (R.ConcreteAddress MA.ARM -> R.SymbolicAddress MA.ARM)
+                      -> MD.ParsedBlock MA.ARM ids
                       -> R.ConcreteAddress MA.ARM
-                      -> Maybe (R.SymbolicAddress MA.ARM)
                       -> Instruction tp ()
-                      -> [R.TaggedInstruction MA.ARM tp TargetAddress]
-armSymbolizeAddresses _mem insnAddr mSymbolicTarget i =
+                      -> [R.Instruction MA.ARM tp (R.Relocation MA.ARM)]
+armSymbolizeAddresses mem toSymbolic pb insnAddr i =
   case i of
-    ARMInstruction (armDropAnnotations -> DA.Instruction opc operands) ->
-      case opc of
-        DA.LDR_l_A1 ->
-          case operands of
-            -- LDRT_A1 :: Opcode o '["Bv1", "Bv4", "Bv1", "Bv1", "Bv4", "Bv12"]
-            --
-            -- P, Rt, U, W, cond, imm12
-            --
-            -- NOTE: So far, all PC-relative address computation in gcc-compiled
-            -- binaries seems to be done using this instruction
-            p DA.:< rt DA.:< u DA.:< w DA.:< cond DA.:< op12@(DA.Bv12 (asInteger -> off12)) DA.:< DA.Nil ->
-              let target = insnAddr `R.addressAddOffset` fromIntegral off12
-                  i' = DA.Instruction (coerce opc) (     noAddr p
-                                                   DA.:< noAddr rt
-                                                   DA.:< noAddr u
-                                                   DA.:< noAddr w
-                                                   DA.:< noAddr cond
-                                                   DA.:< withAddr target op12
+    ARMInstruction (DA.Instruction opc operands) ->
+      case (opc, operands) of
+        (DA.LDR_l_A1, DA.Annotated _ p
+                DA.:< DA.Annotated _ rt
+                DA.:< DA.Annotated _ u
+                DA.:< DA.Annotated _ w
+                DA.:< DA.Annotated _ cond
+                DA.:< DA.Annotated _ (DA.Bv12 (asInteger -> off12))
+                DA.:< DA.Nil) ->
+          -- See Note [Rewriting LDR] for details on this construction
```

Comment:
This formatting is a lot easier to read than the one-line version used in `armConcretizeAddresses` above.

Reply:
Updated